### PR TITLE
Sapphire - Ariel Williams 

### PR DESCRIPTION
**Pull request by arielwilliams**

### CODE REVIEW

**Comment by arielwilliams** [comment]
Completed wave 5 so project is now completed.

File: tests/test_wave_01.py [hunk 1]
```diff
@@ -152,19 +152,32 @@ def test_moves_movie_from_watchlist_to_empty_watched():
...
+    movie_in_index_0 = janes_data["watchlist"][0]
+
     # Act
     updated_data = watch_movie(janes_data, MOVIE_TITLE_1)
 
     # Assert
     assert len(updated_data["watchlist"]) == 0
     assert len(updated_data["watched"]) == 1
+    # modifications added assert statement below 
+    assert movie_in_index_0 in updated_data["watched"]
```
Comment (audreyandoy) [inline tests/test_wave_01.py:165]: 👍

File: tests/test_wave_01.py [hunk 2]
```diff
@@ -182,13 +195,21 @@ def test_moves_movie_from_watchlist_to_watched():
...
     assert len(updated_data["watchlist"]) == 1
     assert len(updated_data["watched"]) == 2
-    
-    raise Exception("Test needs to be completed.")
+    # assert movie_to_watch in updated_data["watched"]
+    ##### OR  assert movie_to_watch in updated_data["watchlist"] #######
+    # modifications added three assert statements below
+    assert HORROR_1 in updated_data["watched"]
+    assert FANTASY_2 in updated_data["watched"]
+    assert FANTASY_1 in updated_data["watchlist"]
```
Comment (audreyandoy) [inline tests/test_wave_01.py:203]: 👍

File: tests/test_wave_03.py [hunk 3]
```diff
@@ -43,24 +43,25 @@ def test_friends_unique_movies():
...
-
+    expected = [FANTASY_4, HORROR_1, INTRIGUE_3]
     # Act
     friends_unique_movies = get_friends_unique_watched(amandas_data)
 
     # Assert
     assert len(friends_unique_movies) == 3
-
-    raise Exception("Test needs to be completed.")
+    assert expected == friends_unique_movies
```
Comment (audreyandoy) [inline tests/test_wave_03.py:57]: 👍

File: tests/test_wave_05.py [hunk 4]
```diff
@@ -53,12 +53,19 @@ def test_new_genre_rec_from_empty_friends():
         ]
     }
 
-    raise Exception("Test needs to be completed.")
+    
+    # Act # (I added this)
+    recommendations = get_new_rec_by_genre(sonyas_data)
+
+    # Assert (I added this)
+    assert len(recommendations) == 0
```
Comment (audreyandoy) [inline tests/test_wave_05.py:61]: 👍

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,217 @@
...
+    output: dictionary with the three keys "title", "genre", and "rating"
+    """
+    movie_dict = {}
+    if title and genre and rating:
+        movie_dict["title"] = title
+        movie_dict["genre"] = genre
+        movie_dict["rating"] = rating
+        return movie_dict
+
+    return None
```
Comment (audreyandoy) [inline viewing_party/party.py:18]: 👍 Great use of a docstring! We can use a guard clause to exit the function early and return the dictionary literal itself to save 1 line (a really minor nitpick). 

```python
def create_movie(title, genre, rating):
    # Input validation
    if not title or not genre or not rating:
        return None

    return {
        "title": title,
        "genre": genre,
        "rating": rating
    }
```

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,217 @@
...
+
+def add_to_watchlist(user_data, movie):
+    """
+    input: user_data dictionary, and movie dictionary
+    output: updated user_data dictionary
+    """
+
+    user_data["watchlist"].append(movie) 
+
+    return user_data
```
Comment (audreyandoy) [inline viewing_party/party.py:40]: 👍 Nice work! 

Here's some food for thought:

If the movie being passed in isn't valid (missing keys, wrong data type, etc.)? A helper function to sanitize the inputs would be helpful here!

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,217 @@
...
+    output: updated user_data dictionary 
+    """ 
+
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+            return user_data
+    
+    return user_data
```
Comment (audreyandoy) [inline viewing_party/party.py:55]: 👍

Comment (audreyandoy) [reply viewing_party/party.py:55]: We could use `break` within the conditional just to slightly DRY up this code:

```python
for movie in user_data["watchlist"]:
        if movie["title"] == title:
            user_data["watchlist"].remove(movie)
            user_data["watched"].append(movie)
            break

    return user_data
```

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,217 @@
...
+        rating = watched["rating"]
+        movie_ratings.append(rating)
+    
+    sum_of_movie_ratings = sum(movie_ratings)
+    length_of_movie_ratings = len(movie_ratings)
+    if length_of_movie_ratings == 0:
+        return average_rating
+    average_rating = sum_of_movie_ratings/length_of_movie_ratings
+
+    return average_rating
```
Comment (audreyandoy) [inline viewing_party/party.py:81]: 👍  Nice work!!

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,217 @@
...
+    movie_genres = []
+
+    for watched in user_data["watched"]:
+        genre = watched["genre"]
+        movie_genres.append(genre)
+    if len(movie_genres) == 0:
+        return None
+
+    most_frequently_watched_genre = (mode(movie_genres))
+    return most_frequently_watched_genre
```
Comment (audreyandoy) [inline viewing_party/party.py:100]: 👍

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,217 @@
...
+    friends_watched_list = []
+    movies_user_watched_that_friends_didnt_watch = []
+    
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_list.append(movie)
+
+    for user in user_data["watched"]:
+        user_watched_list.append(user)
```
Comment (audreyandoy) [inline viewing_party/party.py:123]: We can use list comprehension to build `user_watched_list`:

```python
user_watched_list = [for user in user_data["watched"]]
```

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,217 @@
...
+
+
+    for movie in user_watched_list: 
+        if movie not in friends_watched_list:
+            movies_user_watched_that_friends_didnt_watch.append(movie)
```
Comment (audreyandoy) [inline viewing_party/party.py:128]: List comprehension can also be used here too:

```python

movies_user_watched_that_friends_didnt_watch = [ movie for movie in user_watched_list if movie not in friends_watched_list ]


```

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,217 @@
...
+    """
+
+    user_watched_list = []
+    friends_watched_list = []
+    movies_user_watched_that_friends_didnt_watch = []
+    
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_list.append(movie)
```
Comment (audreyandoy) [inline viewing_party/party.py:120]: We can move this piece into a helper function as the next few functions need this same logic of "flattening" the friends' movies into a single list.

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,217 @@
...
+
+    for user in user_data["watched"]:
+        user_watched_list.append(user)
+
+    for movie in friends_watched_list:
+        if movie not in user_watched_list and movie not in movies_friends_watched_that_user_didnt_watch:
+            movies_friends_watched_that_user_didnt_watch.append(movie)
+
+    return movies_friends_watched_that_user_didnt_watch
```
Comment (audreyandoy) [inline viewing_party/party.py:154]: Similar comments from `get_unique_watched` can be applied to this function (using list comprehension and helper function for the friends' movies list).

File: viewing_party/party.py [hunk 14]
```diff
@@ -1,23 +1,217 @@
...
+    input: user_data dictionary with field "subscriptions", value of "subscriptions" is list of strings
+    output: recommended movies list
+    """
+    rec_movies = []
+    unique_list = get_friends_unique_watched(user_data)
+
+    for movie in unique_list:
+        if movie["host"] in user_data["subscriptions"]:
+            rec_movies.append(movie)
+    return rec_movies
```
Comment (audreyandoy) [inline viewing_party/party.py:172]: 👍 Yay helper functions!!

File: viewing_party/party.py [hunk 15]
```diff
@@ -1,23 +1,217 @@
...
+    users_most_frequent_genre = get_most_watched_genre(user_data)
+
+    if len(users_watched) == 0 or len(users_watched) == 0:
+        return recommendation_by_genre
+    
+    for movie in friends_watched:
+        if movie["genre"] == users_most_frequent_genre:
+            recommendation_by_genre.append(movie)
+
+    return recommendation_by_genre
```
Comment (audreyandoy) [inline viewing_party/party.py:198]: 👍

File: viewing_party/party.py [hunk 16]
```diff
@@ -1,23 +1,217 @@
...
+    recommended_movies_list = []
+
+    users_unique_movies = get_unique_watched(user_data)
+    users_favorite_movies = user_data["favorites"]
+
+    for movie in users_unique_movies:
+        if movie in users_favorite_movies:
+            recommended_movies_list.append(movie)
+
+    return recommended_movies_list
```
Comment (audreyandoy) [inline viewing_party/party.py:216]: 👍

**Review by audreyandoy — COMMENTED** [review]
Great work Ariel and Monica! You hit all the learning goals for this project and all tests are passing. You have earned a well-deserved 🟢 grade on this project ✨

I added comments, compliments, and hints on ways to refactor your code.

Keep up the great work! ✨

**Pull request closed without merging**